**Sprint 10 -- Silvia Vilches**

---

# **NIVEL 1: APIs REST**


---

🌍 Desde un Jupyter Notebook, realizarás los siguientes ejercicios utilizando la biblioteca «requests» de Python.

Exploración básica de una API de laboratorio.
1. Consulta la API pública de JSONPlaceholder utilizando el método GET para obtener:

* Lista de publicaciones (/posts)
* Lista de usuarios (/users)
* Lista de tareas (/todos)

2. Muestra en pantalla:

* El recuento total de cada recurso.
* El código de estado de cada solicitud.

3. Realiza una solicitud a una publicación inexistente para obtener un error 404 y muestra el código de estado recibido.


4. Realiza una solicitud POST para crear una nueva publicación ficticia. Incluye el título, el cuerpo del mensaje y un userId.

* Mostrar la respuesta JSON.
* Mostrar el código de estado.

5. Realizar una solicitud PATCH para modificar parcialmente una entrada existente.

* Mostrar la respuesta JSON.
* Mostrar el código de estado.

6. Realizar una solicitud DELETE sobre una entrada.

* Mostrar la respuesta JSON.
* Mostrar el código de estado.

---

**>> Instalamos la libreria request**

Necesaria para realizar las solicitudes HTTP que nos piden.

        
        pip install requests

In [143]:
import requests

**>> Recurso para realizar las consultas**

        https://jsonplaceholder.typicode.com/

**1. Consulta la API pública de JSONPlaceholder utilizando el método GET para obtener:**

* Lista de publicaciones (/posts)

In [144]:
request_post = requests.get("https://jsonplaceholder.typicode.com/posts")

print(request_post)

<Response [200]>


💠 El código 200 os dice que todo está “OK” ➡ La respuesta para una solicitud HTTP ha tenido éxito.

In [145]:
posts = request_post.json()
print(f"Lista de publicaciones:{posts}")

Lista de publicaciones:[{'userId': 1, 'id': 1, 'title': 'sunt aut facere repellat provident occaecati excepturi optio reprehenderit', 'body': 'quia et suscipit\nsuscipit recusandae consequuntur expedita et cum\nreprehenderit molestiae ut ut quas totam\nnostrum rerum est autem sunt rem eveniet architecto'}, {'userId': 1, 'id': 2, 'title': 'qui est esse', 'body': 'est rerum tempore vitae\nsequi sint nihil reprehenderit dolor beatae ea dolores neque\nfugiat blanditiis voluptate porro vel nihil molestiae ut reiciendis\nqui aperiam non debitis possimus qui neque nisi nulla'}, {'userId': 1, 'id': 3, 'title': 'ea molestias quasi exercitationem repellat qui ipsa sit aut', 'body': 'et iusto sed quo iure\nvoluptatem occaecati omnis eligendi aut ad\nvoluptatem doloribus vel accusantium quis pariatur\nmolestiae porro eius odio et labore et velit aut'}, {'userId': 1, 'id': 4, 'title': 'eum et est occaecati', 'body': 'ullam et saepe reiciendis voluptatem adipisci\nsit amet autem assumenda provident 

💠 Nos devuelve la lista con las publicaciones que hay en la web *https://jsonplaceholder.typicode.com*

* Lista de usuarios (/users)

In [146]:
request_users = requests.get("https://jsonplaceholder.typicode.com/users")

print(request_users)

<Response [200]>


In [147]:
users = request_users.json()
print(f"Lista de usuarios:{users}")

Lista de usuarios:[{'id': 1, 'name': 'Leanne Graham', 'username': 'Bret', 'email': 'Sincere@april.biz', 'address': {'street': 'Kulas Light', 'suite': 'Apt. 556', 'city': 'Gwenborough', 'zipcode': '92998-3874', 'geo': {'lat': '-37.3159', 'lng': '81.1496'}}, 'phone': '1-770-736-8031 x56442', 'website': 'hildegard.org', 'company': {'name': 'Romaguera-Crona', 'catchPhrase': 'Multi-layered client-server neural-net', 'bs': 'harness real-time e-markets'}}, {'id': 2, 'name': 'Ervin Howell', 'username': 'Antonette', 'email': 'Shanna@melissa.tv', 'address': {'street': 'Victor Plains', 'suite': 'Suite 879', 'city': 'Wisokyburgh', 'zipcode': '90566-7771', 'geo': {'lat': '-43.9509', 'lng': '-34.4618'}}, 'phone': '010-692-6593 x09125', 'website': 'anastasia.net', 'company': {'name': 'Deckow-Crist', 'catchPhrase': 'Proactive didactic contingency', 'bs': 'synergize scalable supply-chains'}}, {'id': 3, 'name': 'Clementine Bauch', 'username': 'Samantha', 'email': 'Nathan@yesenia.net', 'address': {'stree

* Lista de tareas (/todos)

In [148]:
request_todos = requests.get("https://jsonplaceholder.typicode.com/todos")

print(request_todos)

<Response [200]>


In [149]:
todos = request_post.json()
print(f"La lista completa:{todos}")

La lista completa:[{'userId': 1, 'id': 1, 'title': 'sunt aut facere repellat provident occaecati excepturi optio reprehenderit', 'body': 'quia et suscipit\nsuscipit recusandae consequuntur expedita et cum\nreprehenderit molestiae ut ut quas totam\nnostrum rerum est autem sunt rem eveniet architecto'}, {'userId': 1, 'id': 2, 'title': 'qui est esse', 'body': 'est rerum tempore vitae\nsequi sint nihil reprehenderit dolor beatae ea dolores neque\nfugiat blanditiis voluptate porro vel nihil molestiae ut reiciendis\nqui aperiam non debitis possimus qui neque nisi nulla'}, {'userId': 1, 'id': 3, 'title': 'ea molestias quasi exercitationem repellat qui ipsa sit aut', 'body': 'et iusto sed quo iure\nvoluptatem occaecati omnis eligendi aut ad\nvoluptatem doloribus vel accusantium quis pariatur\nmolestiae porro eius odio et labore et velit aut'}, {'userId': 1, 'id': 4, 'title': 'eum et est occaecati', 'body': 'ullam et saepe reiciendis voluptatem adipisci\nsit amet autem assumenda provident rerum

**2. Muestra en pantalla:**

* El recuento total de cada recurso.
* El código de estado de cada solicitud.


In [150]:
print(f"Número total de publicaciones: {len(posts)}")
print(f"Código de estado de la solicitud de posts: {request_post.status_code}\n")

print(f"Número total de usuarios: {len(users)}")
print(f"Código de estado de la solicitud de users: {request_users.status_code}\n")

print(f"Número total: {len(todos)}")
print(f"Código de estado de la solicitud: {request_todos.status_code}")

Número total de publicaciones: 100
Código de estado de la solicitud de posts: 200

Número total de usuarios: 10
Código de estado de la solicitud de users: 200

Número total: 100
Código de estado de la solicitud: 200


💠 El código 200 nos dice que todo está “OK” ➡ La respuesta para una solicitud HTTP ha tenido éxito.

💠 Como los datos que nos devuelve la petición son listas, contamos cuantos elementos tiene la lista ➡ *len*

**3. Realiza una solicitud a una publicación inexistente para obtener un error 404 y muestra el código de estado recibido.**


In [151]:
cod_404 = requests.get(f"https://jsonplaceholder.typicode.com/posts/posts/567")

print(f"Código de estado recibido: {cod_404.status_code}")

Código de estado recibido: 404


💠 Como sabemos que hay 100 post, buscamos cualquier que sea superior a este número.

💠 El código 404 nos dice que no ha encontrado lo que buscamos ➡ La URL que queremos ver no existe.

**4. Realiza una solicitud POST para crear una nueva publicación ficticia. Incluye el título, el cuerpo del mensaje y un userId.**

* Mostrar la respuesta JSON.
* Mostrar el código de estado.




In [152]:
new_post = {
    "title": "Hola, esta es mi publicación falsa.",
    "body": "El veloz murciélago hindú comía feliz cardillo y kiwi.",
    "userId": 26
}

resp_post = requests.post(f"https://jsonplaceholder.typicode.com/posts", json=new_post)

print("Respuesta JSON:", resp_post.json())
print("Código de estado:", resp_post.status_code)


Respuesta JSON: {'title': 'Hola, esta es mi publicación falsa.', 'body': 'El veloz murciélago hindú comía feliz cardillo y kiwi.', 'userId': 26, 'id': 101}
Código de estado: 201


💠Hemos creado una publicación donde seguimos el formato del diccionario.

💠 EL código 201 nos indica que el post se ha creado correctamente ➡ Creado con éxito.

**5. Realizar una solicitud PATCH para modificar parcialmente una entrada existente.**

* Mostrar la respuesta JSON.
* Mostrar el código de estado.

In [153]:
update_post = {
    "title": "UPDATE PATCH"
}

resp_patch = requests.patch(f"https://jsonplaceholder.typicode.com/posts/4", json=update_post)

print("Respuesta JSON:", resp_patch.json())
print("Código de estado:", resp_patch.status_code)


Respuesta JSON: {'userId': 1, 'id': 4, 'title': 'UPDATE PATCH', 'body': 'ullam et saepe reiciendis voluptatem adipisci\nsit amet autem assumenda provident rerum culpa\nquis hic commodi nesciunt rem tenetur doloremque ipsam iure\nquis sunt voluptatem rerum illo velit'}
Código de estado: 200


💠 El código 200 nos dice que todo está “OK” ➡ La respuesta para una solicitud HTTP ha tenido éxito.

💠 Hemos modificado el titulo de la publicación con ID 4.

**6. Realizar una solicitud DELETE sobre una entrada.**

* Mostrar la respuesta JSON.
* Mostrar el código de estado.

In [154]:
resp_delete = requests.delete(f"https://jsonplaceholder.typicode.com/posts/4")

print("Respuesta JSON:", resp_delete.json())
print("Código de estado:", resp_delete.status_code)

Respuesta JSON: {}
Código de estado: 200


💠 El código 200 nos dice que todo está “OK” ➡ La respuesta para una solicitud HTTP ha tenido éxito.

💠 Eliminamos la entrada quehabiamos hecho.

---

# **NIVEL 2: Interacción con una API pública real**


---

1. Explora el repositorio de API públicas y elige una API que permita solicitudes GET.


2. Lee la documentación de la API:

* Revisa la sección sobre los puntos finales disponibles. Anota en un archivo Markdown al menos dos puntos finales diferentes.
* Comprueba si ofrece algún filtro o parámetro opcional interesante. Anótalos en un archivo Markdown.
* Verifica que la respuesta esté en formato JSON (es el más adecuado para este ejercicio).

3. Realiza una solicitud GET sencilla:

* Muestra el código de estado de la respuesta.
* Imprime de forma clara algunos de los campos de la respuesta JSON.

4. Convierte los datos a un DataFrame de pandas:

* Muestra las primeras filas del DataFrame.

---

In [155]:
import requests
import pandas as pd

**1. Explora el repositorio de API públicas y elige una API que permita solicitudes GET.**

                https://github.com/public-apis/public-apis

💠 He elegido una APl de Anime/Entretenimiento porque puedo hace solicitudes **GET** sin restricciones y las respuestas están en formato *JSON*.

                 https://pokeapi.co/api/v2/


In [156]:
pokeAPI = "https://pokeapi.co/api/v2/"

**2. Lee la documentación de la API:**

* Revisa la sección sobre los puntos finales disponibles. Anota en un archivo Markdown al menos dos puntos finales diferentes.
* Comprueba si ofrece algún filtro o parámetro opcional interesante. Anótalos en un archivo Markdown.
* Verifica que la respuesta esté en formato JSON (es el más adecuado para este ejercicio).

---

**Puntos finales disponibles** :
* `GET /pokemon` → listado paginado de Pokémon (nombre + URL de detalle).
* `GET /pokemon/id` → información completa del Pokémon.


**Parámetros opcionales interesantes** del endpoint */pokemon*:
*  `limit`: número máximo de resultados que queremos.
*  `offset`: desde que id de pokémon queremos empezar

**La respuesta se devuelve en formato JSON.**

In [167]:
{
  "count": 1302,
  "next": "https://pokeapi.co/api/v2/pokemon?offset=20&limit=20",
  "results": [
    { "name": "bulbasaur", "url": "https://pokeapi.co/api/v2/pokemon/1/" },
    { "name": "ivysaur", "url": "https://pokeapi.co/api/v2/pokemon/2/" }
  ]
}

{'count': 1302,
 'next': 'https://pokeapi.co/api/v2/pokemon?offset=20&limit=20',
 'results': [{'name': 'bulbasaur',
   'url': 'https://pokeapi.co/api/v2/pokemon/1/'},
  {'name': 'ivysaur', 'url': 'https://pokeapi.co/api/v2/pokemon/2/'}]}

**3. Realiza una solicitud GET sencilla:**

* Muestra el código de estado de la respuesta.
* Imprime de forma clara algunos de los campos de la respuesta JSON.


💠Hacemos una petición GET, y comprobamos el código de estado y el contenido de la respuesta.

In [178]:
params = {
    "limit": 20,   # número de resultados que queremos
    "offset": 0    # desde qué posición empezar
}

respuesta_poke = requests.get(f"{pokeAPI}/pokemon", params=params)


In [179]:
print(f"El código de estado de esta petición es: {respuesta_poke.status_code}")

El código de estado de esta petición es: 200


💠Convertimos la respuesta JSON a un diccionario de Python

In [185]:
datos_poke = respuesta_poke.json()

💠Mostramos algunos campos del JSON.

In [186]:
print("Número total de Pokémon:", datos_poke['count'])
print("URL de la siguiente página:", datos_poke['next'])


Número total de Pokémon: 1351
URL de la siguiente página: https://pokeapi.co/api/v2/pokemon?offset=20&limit=20


**4. Convierte los datos a un DataFrame de pandas:**

* Muestra las primeras filas del DataFrame.

In [188]:
df_pokemon = pd.DataFrame(datos_poke["results"])

print("Forma del DataFrame (filas, columnas):", df_pokemon.shape)
df_pokemon.head(5)

Forma del DataFrame (filas, columnas): (20, 2)


,name,url
0,bulbasaur,https://pokeapi.co/api/v2/pokemon/1/
1,ivysaur,https://pokeapi.co/api/v2/pokemon/2/
2,venusaur,https://pokeapi.co/api/v2/pokemon/3/
3,charmander,https://pokeapi.co/api/v2/pokemon/4/
4,charmeleon,https://pokeapi.co/api/v2/pokemon/5/


---

# **NIVEL 3: API de datos abiertos de Barcelona**

        https://opendata-ajuntament.barcelona.cat/ca/desenvolupadors#APIS

---

1. Utiliza la API de datos abiertos de Barcelona para:

* Buscar un conjunto de datos que te interese utilizando package_search y package_show.

2. De entre los resultados obtenidos, selecciona uno que tenga recursos disponibles en formato CSV o JSON.


3. Utilizando el resource_id del recurso seleccionado, realiza una consulta con datastore_search para:

* Recuperar al menos 100 registros.

4. Convierte los resultados a un DataFrame.


5. Guarda el DataFrame en un archivo .csv.


---

In [ ]:
ajuntamentBCN = "https://opendata-ajuntament.barcelona.cat/data/api"

**1. Utiliza la API de datos abiertos de Barcelona para:**

* Buscar un conjunto de datos que te interese utilizando package_search y package_show.

💠 Buscamos datasets relacionados con el arbolado ("arbrat") de Barcelona.

In [193]:
params = {"q": "arbrat", "rows": 10} # la letra q --> query, consulta.

resp_busqueda = requests.get(f"{ajuntamentBCN}/3/action/package_search", params=params)

print("Código de estado:", resp_busqueda.status_code)

resultado_busqueda = resp_busqueda.json()['result']
print(f"Número de datasets encontrados: {resultado_busqueda['count']}")
print()
for ds in resultado_busqueda['results']:
    print(f" - {ds['name']}  |  {ds['title']}")

Código de estado: 200
Número de datasets encontrados: 7

 - atles-biodiversitat-arbrat-rel-geometria-especie  |  Geometry-species relationship of the trees of the Biodiversity Atlas of the city of Barcelona
 - atles-biodiversitat-arbrat-geometries  |  Inventory of tree geometries of the Biodiversity Atlas of the city of Barcelona
 - atles-biodiversitat-arbrat-especies  |  Inventory of tree species of the Atlas of Biodiversity of the city of Barcelona
 - arbrat-parcs  |  Trees in the parks of the city of Barcelona
 - arbrat-zona  |  Zone trees of the city of Barcelona
 - arbres-interes-local  |  Local interest trees from Barcelona
 - arbrat-viari  |  Street trees of the city of Barcelona


**2. De entre los resultados obtenidos, selecciona uno que tenga recursos disponibles en formato CSV o JSON.**

💠 Elegimos el dataset *arbrat-zona* (árboles de zona de Barcelona) y consultamos su detalle para ver qué recursos tiene disponibles y cuáles permiten consulta vía *datastore_search*.

In [198]:
resp_arboles = requests.get(f"{ajuntamentBCN}/3/action/package_show", params={"id": "arbrat-zona"})  #consultamos la info, toda la información del conjunto de datos llamado arbrat-zona

print("Código de estado:", resp_arboles.status_code)

data_arboles = resp_arboles.json()["result"]  # convertimos la respuesta en un diccionario.
print("---------------------------------------------------------")
print("Título del daraframe:", data_arboles["title"])
print("---------------------------------------------------------")
print("Recursos con datastore activo:")
for recurso in data_arboles["resources"]:  #buscamos, si el recurso nos permite hacer consultas con datastore_search, que nos muestre la info.
    if recurso.get("datastore_active"):
        print(f"- Nombre: {recurso['name']}")
        print(f"- Formato: {recurso['format']}")
        print(f"- Resource_id: {recurso['id']}")

Código de estado: 200
---------------------------------------------------------
Título del daraframe: Zone trees of the city of Barcelona
---------------------------------------------------------
Recursos con datastore activo:
- Nombre: OD_Arbrat_Zona_BCN.csv
- Formato: CSV
- Resource_id: 29cd5c1f-11b1-404b-b3a5-ae29940b8c55


In [199]:
# Nos quedamos con el recurso CSV, que es el que tiene el datastore activo, que nos ha dado en la consulta anterior
resource_id = "29cd5c1f-11b1-404b-b3a5-ae29940b8c55" 
print("resource_id:", resource_id)

resource_id: 29cd5c1f-11b1-404b-b3a5-ae29940b8c55


**3. Utilizando el resource_id del recurso seleccionado, realiza una consulta con datastore_search para:**

* Recuperar al menos 100 registros.

In [196]:
params = {
    "resource_id": resource_id,
    "limit": 100
}

respuesta_ajt = requests.get(f"{ajuntamentBCN}/action/datastore_search", params=params)

print("Código de estado:", respuesta_ajt.status_code)

data = respuesta_ajt.json()['result']
print("---------------------------------------------------------")
print("Campos disponibles:", [f["id"] for f in data["fields"]])
print("---------------------------------------------------------")
print("Número de registros recuperados:", len(data["records"]))

Código de estado: 200
---------------------------------------------------------
Campos disponibles: ['_id', 'codi', 'x_etrs89', 'y_etrs89', 'latitud', 'longitud', 'tipus_element', 'espai_verd', 'adreca', 'cat_especie_id', 'cat_nom_cientific', 'cat_nom_castella', 'cat_nom_catala', 'categoria_arbrat', 'data_plantacio', 'tipus_aigua', 'tipus_reg', 'geom', 'catalogacio', 'codi_barri', 'nom_barri', 'codi_districte', 'nom_districte']
---------------------------------------------------------
Número de registros recuperados: 100


**4. Convierte los resultados a un DataFrame.**


In [ ]:
df_arboles = pd.DataFrame(data['records'])

print("Forma del DataFrame:", df_arboles.shape)
df_arboles.head(5)

Forma del DataFrame: (100, 23)


,latitud,cat_nom_catala,codi,espai_verd,adreca,catalogacio,nom_districte,tipus_element,cat_nom_castella,nom_barri,...,data_plantacio,codi_barri,longitud,codi_districte,tipus_reg,categoria_arbrat,cat_especie_id,tipus_aigua,y_etrs89,_id
0,41.4365081,Palmera datilera; palmera de dàtils,0000377AR,"Central de Nou Barris, Parc","Pl Major de Nou Barris, 0",,NOU BARRIS,PALMERA ZONA,Palmera datilera,LA GUINEUETA,...,NaN,48,2.1705107,8,GOTEIG,TERCERA,48,,4587547.432,1
1,41.4365615,palmera de la reina,0000378AR,"Central de Nou Barris, Parc","Pl Major de Nou Barris, 0",,NOU BARRIS,PALMERA ZONA,palmera pindó,LA GUINEUETA,...,2026-06-10T00:00:00,48,2.1705276,8,GOTEIG,PRIMERA,485,,4587553.347,2
2,41.4377690,Plàtan,0000386AR,"Central de Nou Barris, Parc","C\ Doctor Letamendi, 90",,HORTA - GUINARDÓ,ARBRE ZONA,Plátano,HORTA,...,NaN,43,2.1625299,7,SENSE INFORMAR,SEGONA,126,,4587693.836,3
3,41.4377883,Plàtan,0000387AR,"Central de Nou Barris, Parc","C\ Doctor Letamendi, 90",,HORTA - GUINARDÓ,ARBRE ZONA,Plátano,HORTA,...,NaN,43,2.1625008,7,SENSE INFORMAR,SEGONA,126,,4587696.005,4
4,41.4378096,Plàtan,0000388AR,"Central de Nou Barris, Parc","C\ Doctor Letamendi, 90",,HORTA - GUINARDÓ,ARBRE ZONA,Plátano,HORTA,...,NaN,43,2.1624711,7,SENSE INFORMAR,TERCERA,126,,4587698.393,5


**5. Guarda el DataFrame en un archivo .csv.**

In [ ]:
df_arboles.to_csv("arbolest_zona_barcelona.csv", index=False, encoding="utf-8-sig")
print("Archivo guardado como - arbolest_zona_barcelona.csv - ")

Archivo guardado como - arbolest_zona_barcelona.csv - 
